### PR TITLE
[MM-21369] Require server/site URL for deep links

### PR DESCRIPTION
**Pull request by chuttam**
#### Summary

Deep links now need to have the server (or site) URL as a part of the request. An absent or incorrect URL throws up an alert popup stating that the link was incorrect.

**Old (only worked without serverURL or siteURL)**

Beta
* `mattermost-beta://<teamname>/channels/<channelname>`
* `mattermost-beta://<teamname>/pl/<permalinkID>`

Release
* `mattermost-mobile://<teamname>/channels/<channelname>`
* `mattermost-mobile://<teamname>/pl/<permalinkID>`

**New working deep link patterns**

Note: Transport protocol (http, https, etc.) is optional.

Beta
* `mattermost-beta://<server-or-site-URL><teamname>/channels/<channelname>`
* `mattermost-beta://<server-or-site-URL><teamname>/pl/<permalinkID>`
* `mattermost-beta://https://<server-or-site-URL><teamname>/channels/<channelname>`
* `mattermost-beta://https://<server-or-site-URL><teamname>/pl/<permalinkID>`

Release
* `mattermost-mobile://<server-or-site-URL><teamname>/channels/<channelname>`
* `mattermost-mobile://<server-or-site-URL><teamname>/pl/<permalinkID>`
* `mattermost-mobile://https://<server-or-site-URL><teamname>/channels/<channelname>`
* `mattermost-mobile://https://<server-or-site-URL><teamname>/pl/<permalinkID>`

#### Ticket Link
* [MM-21369](https://mattermost.atlassian.net/browse/MM-21369)

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [x] Includes text changes and localization file updates

#### Device Information
This PR was tested on:
- [x] Android 9 device (OnePlus 5)
- [x] iPhone 11 device

#### Screenshots

| Correct URL| Absent URL | Incorrect URL |
| --- | --- | -- |
| Correct URL | Absent URL | Incorrect URL |

---

**Additional Notes**
* Current error message text is as follows. Needs PM review.

```
Link Error

We were unable to locate a Mattermost server for the link you tried.
Please check the address of the link or contact an administrator.
```

### CODE REVIEW

**Comment by chuttam** [comment]
This PR current merges into branch for PR #3767 , but since review of that PR should be done soon, I felt it best to not merge this PR into that branch.

If #3767 gets merged before this PR is looked at, I'll rebase this branch against `master`.

**Comment by chuttam** [comment]
I also added the same set of reviewers on this PR as on #3767 to make testing a bit quicker, since the setup is the same for both PRs.

**Comment by adamjclarkson** [comment]
What do I need to review on this? Just the error messaging? Or manually end to end test it?

**Comment by chuttam** [comment]
> What do I need to review on this? Just the error messaging? Or manually end to end test it?

Yup, just the error message from Product Review.

File: app/utils/url.js [hunk 1]
```diff
@@ -104,15 +103,18 @@ export function matchDeepLink(url, serverURL, siteURL) {
         return null;
     }
 
-    const linkRoot = `(?:${escapeRegex(APP_SCHEME)}:\\/|${escapeRegex(serverURL)}|${escapeRegex(siteURL)})?`;
+    const serverURLWithoutProtocol = removeProtocol(serverURL);
```
Comment (enahum) [inline app/utils/url.js]: Nice!

**Comment by adamjclarkson** [comment]
How about this for an error message:

`Link Error

We were unable to locate a Mattermost server at the specified address:
__mattermost-beta://10.0.0.1/foo/bar__
`

**Comment by chuttam** [comment]
> How about this for an error message:
> 
> ```
> Link Error
> 
> We were unable to locate a Mattermost server at the specified address:
> mattermost-beta://10.0.0.1/foo/bar
> ```
> 
> Is it possible to show the address that the user entered in the error message?

Yup. However, I want to note that some of these addresses could get really long. And in the case of permalinks, they could look really gnarly. Here's an example:

`http://10.0.0.1/teamname/pl/91dqyaayxfdy3ps6f4nd11b7gw`

The Alert pop-up has a little bit of space to work with, but not too much.

Proceed with your suggested change?

**Comment by enahum** [comment]
Why not something like, `the deep link does not belong to this server` without having to give too much explanation

**Comment by chuttam** [comment]
@adamjclarkson @enahum I condensed the messaging per your suggestions, and here's what we have now when attempting to hit a deep link that's for the current server:

Incorrect server URL

This was discussed with @andrewbrown00 too on Friday, and he mentioned he may have input too.

**Comment by chuttam** [comment]
Android:
https://pr-builds.mattermost.com/mattermost-mobile/build-pr-3770/Mattermost_Beta.apk

iOS:
itms-services://?action=download-manifest&url=https://pr-builds.mattermost.com/mattermost-mobile/build-pr-3770/Mattermost_Beta.plist

**Comment by adamjclarkson** [comment]
OK, this generic message will have to suffice for now. I'd like to be able to be more helpful to users with a more specific explanation and especially some direction for what to do next. But as discussed we'll handle all of that in a new ticket. https://mattermost.atlassian.net/browse/MM-21723.

**Pull request merged**